edk2basetools/VfrCompiler/IfrUtility.py

```diff
@@ -0,0 +1,2238 @@
+
+        self.NewDataType = None
+
+    # two definitions
+    def GetDataTypeSizeByTypeName(self, TypeName):
+        Size = 0
+        pDataType = self.DataTypeList
+        while pDataType != None:
+            if pDataType.TypeName == TypeName:
+                Size = pDataType.TotalSize
+                return Size, VfrReturnCode.VFR_RETURN_SUCCESS
+            pDataType = pDataType.Next
+
+        return Size, VfrReturnCode.VFR_RETURN_UNDEFINED
+
+    def GetDataTypeSizeByDataType(self, DataType):
+        Size = 0
+        DataType = DataType & 0x0F
+        # For user defined data type, the size can't be got by this function.
+        if DataType == EFI_IFR_TYPE_OTHER:
+            return Size, VfrReturnCode.VFR_RETURN_SUCCESS
+        pDataType = self.DataTypeList
+        while pDataType != None:
+            if DataType == pDataType.Type:
+                Size = pDataType.TotalSize
+                return Size, VfrReturnCode.VFR_RETURN_SUCCESS
+            pDataType = pDataType.Next
+
+        return Size, VfrReturnCode.VFR_RETURN_UNDEFINED
+
+    def ExtractStructTypeName(self, VarStr):
+        try:
+            index = VarStr.index(".")
+        except ValueError:
+            return VarStr, len(VarStr)
+        else:
+            return VarStr[0:index], index + 1
+
+    def ExtractFieldNameAndArrary(self, VarStr, s):
+        ArrayIdx = INVALID_ARRAY_INDEX
+        s_copy = s
+        while s < len(VarStr) and VarStr[s] != "." and VarStr[s] != "[" and VarStr[s] != "]":
+            s += 1
+
+        FName = VarStr[s_copy:s]
+
+        if s == len(VarStr):
+            return ArrayIdx, s, FName, VfrReturnCode.VFR_RETURN_SUCCESS
+
+        elif VarStr[s] == ".":
+            s += 1
+            return ArrayIdx, s, FName, VfrReturnCode.VFR_RETURN_SUCCESS
+
+        elif VarStr[s] == "[":
+            s += 1
+            try:
+                e = s + VarStr[s:].index("]")
+
+            except ValueError:
+                return None, None, None, VfrReturnCode.VFR_RETURN_DATA_STRING_ERROR
+            else:
+                ArrayStr = VarStr[s:e]
+                ArrayIdx = int(ArrayStr)
+                if VarStr[e] == "]":
+                    e += 1
+                if e < len(VarStr) and (VarStr[e] == "."):
+                    e += 1
+                return ArrayIdx, e, FName, VfrReturnCode.VFR_RETURN_SUCCESS
+
+        elif VarStr[s] == "]":
+            return None, None, None, VfrReturnCode.VFR_RETURN_DATA_STRING_ERROR
+
+        return ArrayIdx, s, FName, VfrReturnCode.VFR_RETURN_SUCCESS
+
+    def GetTypeField(self, FName, Type):
+        if FName == None or Type == None:
+            return None, VfrReturnCode.VFR_RETURN_FATAL_ERROR
+
+        pField = Type.Members
+        while pField != None:
+            if Type.Type == EFI_IFR_TYPE_TIME:
+                if FName == "Hour":
+                    FName = "Hours"
+                elif FName == "Minute":
+                    FName == "Minutes"
```

Comment:
This is wrong. I'm guessing you meant to set this equal to FName, not do another comparison.

Reply:
Hi, here is to change the `FName` from Singular to Plural, and the original C code in **edk2\BaseTools\Source\C\VfrCompile\VfrUtilityLib.cpp** is as follows, and the python codes aim to do the same thing here.
```
if (Type->mType == EFI_IFR_TYPE_TIME) {
if (strcmp (FName, "Hour") == 0) {
FName = "Hours";
} else if (strcmp (FName, "Minute") == 0) {
FName = "Minuts";
} else if (strcmp (FName, "Second") == 0) {
FName = "Seconds";
}
}
```